### ISSUE TITLE
Soft Exponential activation function

### ISSUE THREAD

**Issue by hobson** [fchollet/keras#3842]
PR #3851 

[This paper](https://arxiv.org/pdf/1602.01321.pdf) explains the advantages of a  _soft exponential_ activation function including improved training, simplified neural nets, and improved accuracy for models that rely on multiplicative transforms of the input, such as dot product, cross product, and polynomial transforms.

Draft implementation:

``` python
def softexp(x, alpha=0., max_value=None):
    """Soft Exponential activation function by Godfrey and Gashler

    See: https://arxiv.org/pdf/1602.01321.pdf

    α == 0:  f(α, x) = x
    α  > 0:  f(α, x) = (exp(αx)-1) / α + α
    α  < 0:  f(α, x) = -ln(1-α(x + α)) / α
    """
    if alpha == 0:
        return x
    elif alpha > 0:
        return alpha + (K.exp(alpha * x) - 1.) / alpha
    else:
        return - K.log(1 - alpha * (x + alpha)) / alpha
```

I'm testing it for correctness (match this plot):

![image](https://cloud.githubusercontent.com/assets/292855/18731374/62fc7a44-8010-11e6-8993-0709348d9960.png)

Here's the fork and branch: https://github.com/totalgood/keras/tree/softexp

**Comment by hobson** [comment]
Here's the correctness plot. Looks good, but still needs to verify edge cases.

![image](https://cloud.githubusercontent.com/assets/292855/18733545/bb3eef92-8020-11e6-8db5-325b2dfc8d4f.png)

**Comment by kilotaras** [comment]
Alpha should be trainable. Check ParametricSoftplus implementation

**Comment by hobson** [comment]
Will do. Is there a generalization of TensorFlow and Theano conditionals? Or do I need to build a switch between `tf.cond` and `theano.switch` in `backend.py`?

**Comment by kilotaras** [comment]
`K.switch` should do the trick

**Comment by hobson** [comment]
Perfect. Do you want both a `PararametricSoftExp` *and* a nonparametric `softexp()`?

**Comment by kilotaras** [comment]
I don't think that it makes sense to have nonparametric version. 

I would also recommend getting @fchollet input on this.

**Comment by hobson** [comment]
I attempted to follow the ParametricSoftPlus pattern. But I cannot figure out how to use `K.switch(alpha,...` when `alpha` is trainable (presumably a Tensor), but `switch()` requires a scalar bool as the first arg.  Will send PR so you can comment on lines directly.

```python
class ParametricSoftExp(Layer):
    """Soft Exponential activation function with trainable alpha

    See: https://arxiv.org/pdf/1602.01321.pdf by Godfrey and Gashler

    # Input shape
        Arbitrary. Use the keyword argument `input_shape`
        (tuple of integers, does not include the samples axis)
        when using this layer as the first layer in a model.

    # Output shape
        Same shape as the input.

    # Arguments
        alpha_init: float. Initial value of the alpha weights.
        weights: initial alpha weights, as a list of 1 numpy array.
                 if both weights & alpha_init are provided, weights overrides alpha_init

    Soft Exponential f(α, x):
        α == 0:  x
        α  > 0:  (exp(αx)-1) / α + α
        α  < 0:  -ln(1-α(x + α)) / α
    """
    def __init__(self, alpha_init=0.1,
                 weights=None, **kwargs):
        self.supports_masking = True
        self.alpha_init = K.cast_to_floatx(alpha_init)
        self.initial_weights = weights
        super(ParametricSoftplus, self).__init__(**kwargs)

    def build(self, input_shape):
        input_shape = input_shape[1:]
        self.alphas = K.variable(self.alpha_init * np.ones(input_shape),
                                 name='{}_alphas'.format(self.name))
        self.trainable_weights = [self.alphas]

        if self.initial_weights is not None:
            self.set_weights(self.initial_weights)
            del self.initial_weights  # should this be set to None rather than deleted?

    def call_alpha_gt0(self, x, alpha):
        return alpha + (K.exp(alpha * x) - 1.) / alpha

    def call_alpha_lt0(self, x, alpha):
        return - K.log(1 - alpha * (x + alpha)) / alpha

    def call(self, x, mask=None):
        return K.switch(self.alphas > 0, self.call_alpha_gt0,
                        K.switch(self.alphas < 0, self.call_alpha_lt0, x))

    def get_config(self):
        config = {'alpha_init': float(self.alpha_init)}
        base_config = super(ParametricSoftplus, self).get_config()
        return dict(list(base_config.items()) + list(config.items()))
```

**Comment by zachmayer** [comment]
@hobson I wanted to follow up on this: did you every get a working soft exponential?

**Comment by hobson** [comment]
A 2-input, 1-output, 1 neuron, multiplication model never converged. Don't
know if that's a problem with softexponential itself or my implementation.

--Hobson

**Comment by zachmayer** [comment]
@hobson Could you share the code you ran to test the model?  Maybe the issue is related to https://github.com/fchollet/keras/issues/6621 ?

**Comment by hobson** [comment]
I tried several variations of this "multiply" problem: https://github.com/hobson/toynet/blob/master/toynet/playdot.py

**Comment by hobson** [comment]
@zachmayer I think the problem may be my math. perhaps the multiply problem can only be solved with a layer of SoftExp applied to each input number before summing the weighted outputs with single-neuron linear layer.

**Comment by solalatus** [comment]
Anyone got a stable implementation for this at the end? Great thumbs up for trying!

**Comment by hobson** [comment]
Not me

**Comment by zwep** [comment]
It seems that the ParametricSoftExp given by toynet/hobson is not equal to the one given in the mentioned paper...
From some first tests it seems that with Keras 2.3.0 and Tensorflow 2.0.0 this is a working version


class ParametricSoftExp(klayers.Layer):
    def __init__(self, init_alpha=0.0, init_stdv=0.1, **kwargs):
        """
        Soft Exponential activation function with trainable alpha.
        We initialize alpha from a random uniform distribution.

        Layer can be used as an advanced layer that learns/changes during the optimization process.

       See: https://arxiv.org/pdf/1602.01321.pdf by Godfrey and Gashler

       Soft Exponential f(α, x):
           α == 0:  x
           α  > 0:  (exp(αx)-1) / α + α
           α  < 0:  -ln(1-α(x + α)) / α
        :param init_alpha:
        :param init_stdv:
        :param kwargs:
        """
        super().__init__(**kwargs)
        self.init_mean = init_alpha
        self.init_stdv = init_stdv
        self.atol = tf.constant(1e-08)

    def build(self, temp):
        # Initialize alpha
        alpha_init = tf.random_normal_initializer(mean=self.init_mean, stddev=self.init_stdv)
        self.alpha_actv = tf.Variable(initial_value=alpha_init(shape=(1,), dtype='float32'), trainable=True)

    def call_gt0(self, x):
        return -(tf.math.log(1 - self.alpha_actv * (x + self.alpha_actv))) / (self.alpha_actv)

    def call_lt0(self, x):
        return (tf.math.exp(self.alpha_actv * x) - 1) / self.alpha_actv + self.alpha_actv

    def call(self, x):
        # Check for equal-ness first, based on a certain tolerance.
        cond_equal = tf.less_equal(tf.abs(self.alpha_actv), self.atol)
        # Otherwise go to greater or lower than zero
        res = K.switch(cond_equal, x, K.switch(self.alpha_actv > 0, self.call_gt0(x), self.call_lt0(x)))
        return res

    def get_config(self):
        config = {'alpha_init': float(self.init_mean)}
        base_config = super().get_config()
        return dict(list(base_config.items()) + list(config.items()))


From some simple test it does seem to work really well!

**Comment by hobson** [comment]
Excellent!

**Comment by zwep** [comment]
Just one remark.. In a simple architecture I noticed that the alpha parameter was learning. However, in a more complex model the gradients exploded. Giving only nans as output.. I am still working on that. 
Ideally I want to define my own gradient calculations, but I am not too familiar with Tensorflow yet on how to do that in this case

**Comment by DextroLaev** [comment]
@zwep  hey , I am facing the same problem , I tried to implement it using gradient tape but , the matter is that , sometimes the value inside the ```log()``` becomes a negative as a result it gives nan and we don't get the gradient, can you telll me what's your approach in that case.

**Comment by zachmayer** [comment]
I haven't tried it here, but I've found other places where replacing `log` with `asinh` helps with this problem.

asinh has a very similar shape to the log function, but it is defined at 0 and for negative numbers.

https://worthwhile.typepad.com/worthwhile_canadian_initi/2011/07/a-rant-on-inverse-hyperbolic-sine-transformations.html

**Comment by TheUnsolvedDev** [comment]
The problem after implementing the gradient tape is that the log function becomes undefined under the circumstances of getting a zero value. Any solutions??

**Comment by DextroLaev** [comment]
@TheUnsolvedDev in the forward propagation, it becomes undefined for ```log()``` .

**Comment by TheUnsolvedDev** [comment]
@zachmayer Thanks for the advice I've tried it and got good results. Thank you.

**Comment by EliasHasle** [comment]
> @zwep hey , I am facing the same problem , I tried to implement it using gradient tape but , the matter is that , sometimes the value inside the `log()` becomes a negative as a result it gives nan and we don't get the gradient, can you telll me what's your approach in that case.

I skimmed and Ctrl+F-ed the paper wondering how the function and the authors dealt with negative `x`, but it seems the issue is avoided entirely. Not mentioned once. In light of that, I am not surprised to see that a straightforward implementation trips over this issue...

Is the entire Godfrey&Gashler study trash? Or did they just never have to deal with negative numbers, and so forgot to even think about them? Incredible...

**Comment by shawwn** [comment]
@EliasHasle Hello from the future. We traveled down the same path as you, ran into the same problem as you, and came to the same conclusion as you.

I'm currently trying to think of some way to salvage this idea. But:

> Is the entire Godfrey&Gashler study trash? Or did they just never have to deal with negative numbers, and so forgot to even think about them? Incredible...

I think it's the latter. And yes, it's incredible.

My idea was to let the model curve-fit via a bunch of trainable parameters. No need to only use a single parameter. But it'll be important to initialize it to ReLU so that the model doesn't need to re-learn basic activations from scratch. (Though that raises the problem of, it might stick with whatever activation you initialize it with.)
